`from_over_into` suggests turning `Into` into conflicting `From`

### Summary

For a newtype like `struct Foo(String)`, Rust permits `impl<T> From<T> for Foo where String: From<T>`. In the presence of this, `impl From<Foo> for String` is not allowed due to conflicting implementations, but as a workaround `impl Into<String> for Foo` is permitted. However, `clippy` suggests turning the latter into the former.

### Lint Name

from_over_into

### Reproducer

I tried this code:

```rust
pub struct Foo(pub String);

impl<T> From<T> for Foo
where
    String: From<T>,
{
    fn from(label: T) -> Self {
        Self(String::from(label))
    }
}

// Triggers `clippy::from_over_into`.
impl Into<String> for Foo {
    fn into(self) -> String {
        self.0
    }
}

// Suggestion from `clippy::from_over_into`. Uncommenting gives:
//
// ```
// error[E0119]: conflicting implementations of trait `From<Foo>` for type `Foo`
//  --> src/lib.rs:3:1
//   |
// 3 | / impl<T> From<T> for Foo
// 4 | | where
// 5 | |     String: From<T>,
//   | |____________________^
//   |
//   = note: conflicting implementation in crate `core`:
//           - impl<T> From<T> for T;
// ```
// impl From<Foo> for String {
//     fn from(foo: Foo) -> Self {
//         foo.0
//     }
// }
```

([Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=58f7daf18ca84b0769ba54b77d913f4c))

I saw this happen:

```
warning: an implementation of `From` is preferred since it gives you `Into<_>` for free where the reverse isn't true
  --> src/lib.rs:13:1
   |
13 | impl Into<String> for Foo {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: `impl From<Local> for Foreign` is allowed by the orphan rules, for more information see
           https://doc.rust-lang.org/reference/items/implementations.html#trait-implementation-coherence
   = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.94.0/index.html#from_over_into
   = note: `#[warn(clippy::from_over_into)]` on by default
help: replace the `Into` implementation with `From<Foo>`
   |
13 ~ impl From<Foo> for String {
14 ~     fn from(val: Foo) -> Self {
15 ~         val.0
   |
```

I expected to see this happen: Clippy not suggesting a fix that results in compiler error.


### Version

Tested both stable and latest beta:

```text
rustc 1.95.0-beta.7 (cc969f3c4 2026-04-03)
binary: rustc
commit-hash: cc969f3c457b2ce350fb9fda6c784cc9813b5c71
commit-date: 2026-04-03
host: x86_64-unknown-linux-gnu
release: 1.95.0-beta.7
LLVM version: 22.1.2
```

```text
rustc 1.94.1 (e408947bf 2026-03-25)
binary: rustc
commit-hash: e408947bfd200af42db322daf0fadfe7e26d3bd1
commit-date: 2026-03-25
host: x86_64-unknown-linux-gnu
release: 1.94.1
LLVM version: 21.1.8
```

### Additional Labels

_No response_

<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

--- a/clippy_lints/src/from_over_into.rs
+++ b/clippy_lints/src/from_over_into.rs
@@ -169,9 +169,9 @@ fn has_blanket_from_impl<'tcx>(cx: &LateContext<'tcx>, self_ty: Ty<'tcx>) -> boo
     let Some(from_def_id) = cx.tcx.get_diagnostic_item(sym::From) else {
         return false;
     };
-    cx.tcx.all_impls(from_def_id).any(|impl_id| {
+    cx.tcx.non_blanket_impls_for_ty(from_def_id, self_ty).any(|impl_id| {
         let impl_trait_ref = cx.tcx.impl_trait_ref(impl_id).instantiate_identity();
-        impl_trait_ref.self_ty() == self_ty && matches!(impl_trait_ref.args.type_at(1).kind(), ty::Param(_))
+        matches!(impl_trait_ref.args.type_at(1).kind(), ty::Param(_))
     })
 }
 

--- a/clippy_lints/src/from_over_into.rs
+++ b/clippy_lints/src/from_over_into.rs
@@ -10,11 +10,11 @@ use rustc_errors::Applicability;
 use rustc_hir::intravisit::{Visitor, walk_path};
 use rustc_hir::{
     FnRetTy, GenericArg, GenericArgs, HirId, Impl, ImplItemId, ImplItemKind, Item, ItemKind, PatKind, Path,
-    PathSegment, Ty, TyKind,
+    PathSegment, Ty as HirTy, TyKind,
 };
 use rustc_lint::{LateContext, LateLintPass};
 use rustc_middle::hir::nested_filter::OnlyBodies;
-use rustc_middle::ty;
+use rustc_middle::ty::{self, Ty};
 use rustc_session::impl_lint_pass;
 use rustc_span::symbol::{kw, sym};
 use rustc_span::{Span, Symbol};
@@ -165,7 +165,7 @@ impl<'tcx> Visitor<'tcx> for SelfFinder<'_, 'tcx> {
     }
 }
 
-fn has_blanket_from_impl<'tcx>(cx: &LateContext<'tcx>, self_ty: ty::Ty<'tcx>) -> bool {
+fn has_blanket_from_impl<'tcx>(cx: &LateContext<'tcx>, self_ty: Ty<'tcx>) -> bool {
     let Some(from_def_id) = cx.tcx.get_diagnostic_item(sym::From) else {
         return false;
     };
@@ -178,8 +178,8 @@ fn has_blanket_from_impl<'tcx>(cx: &LateContext<'tcx>, self_ty: ty::Ty<'tcx>) ->
 fn convert_to_from(
     cx: &LateContext<'_>,
     into_trait_seg: &PathSegment<'_>,
-    target_ty: &Ty<'_>,
-    self_ty: &Ty<'_>,
+    target_ty: &HirTy<'_>,
+    self_ty: &HirTy<'_>,
     impl_item_ref: ImplItemId,
 ) -> Option<Vec<(Span, String)>> {
     if !target_ty.find_self_aliases().is_empty() {

--- a/clippy_lints/src/from_over_into.rs
+++ b/clippy_lints/src/from_over_into.rs
@@ -80,6 +80,8 @@ impl<'tcx> LateLintPass<'tcx> for FromOverInto {
             && cx.tcx.is_diagnostic_item(sym::Into, middle_trait_ref.def_id)
             && !matches!(middle_trait_ref.args.type_at(1).kind(), ty::Alias(ty::Opaque, _))
             && self.msrv.meets(cx, msrvs::RE_REBALANCING_COHERENCE)
+            // skip if there's a blanket From impl, the suggested impl would conflict
+            && !has_blanket_from_impl(cx, middle_trait_ref.self_ty())
         {
             span_lint_and_then(
                 cx,
@@ -161,6 +163,16 @@ impl<'tcx> Visitor<'tcx> for SelfFinder<'_, 'tcx> {
             ControlFlow::Continue(())
         }
     }
+}
+
+fn has_blanket_from_impl<'tcx>(cx: &LateContext<'tcx>, self_ty: ty::Ty<'tcx>) -> bool {
+    let Some(from_def_id) = cx.tcx.get_diagnostic_item(sym::From) else {
+        return false;
+    };
+    cx.tcx.all_impls(from_def_id).any(|impl_id| {
+        let impl_trait_ref = cx.tcx.impl_trait_ref(impl_id).instantiate_identity();
+        impl_trait_ref.self_ty() == self_ty && matches!(impl_trait_ref.args.type_at(1).kind(), ty::Param(_))
+    })
 }
 
 fn convert_to_from(

--- a/tests/ui/from_over_into.fixed
+++ b/tests/ui/from_over_into.fixed
@@ -116,4 +116,24 @@ fn issue_112502() {
     }
 }
 
+fn issue_16823() {
+    pub struct Foo(pub String);
+
+    impl<T> From<T> for Foo
+    where
+        String: From<T>,
+    {
+        fn from(val: T) -> Self {
+            Self(String::from(val))
+        }
+    }
+
+    // no lint, From<Foo> for String would conflict with the blanket impl above
+    impl Into<String> for Foo {
+        fn into(self) -> String {
+            self.0
+        }
+    }
+}
+
 fn main() {}

--- a/tests/ui/from_over_into.rs
+++ b/tests/ui/from_over_into.rs
@@ -116,4 +116,24 @@ fn issue_112502() {
     }
 }
 
+fn issue_16823() {
+    pub struct Foo(pub String);
+
+    impl<T> From<T> for Foo
+    where
+        String: From<T>,
+    {
+        fn from(val: T) -> Self {
+            Self(String::from(val))
+        }
+    }
+
+    // no lint, From<Foo> for String would conflict with the blanket impl above
+    impl Into<String> for Foo {
+        fn into(self) -> String {
+            self.0
+        }
+    }
+}
+
 fn main() {}

